warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/edge/ProtobufEdgeDataTypeHandler.java

```diff
@@ -741,6 +609,183 @@ public long process(KEYIN key, RawRecordContainer event, Multimap<String,Normali
+        // get the edgeAttribute3 from the event fields map
+        if (normalizedFields.containsKey(edgeDefConfigs.getEdgeAttribute3())) {
+            return normalizedFields.get(edgeDefConfigs.getEdgeAttribute3()).iterator().next().getIndexedFieldValue();
+        }
+        return null;
+    }
+    
+    private String getEdgeAttr2(EdgeDefinitionConfigurationHelper edgeDefConfigs) {
+        // get the edgeAttribute2 from the event fields map
+        if (normalizedFields.containsKey(edgeDefConfigs.getEdgeAttribute2())) {
+            return normalizedFields.get(edgeDefConfigs.getEdgeAttribute2()).iterator().next().getIndexedFieldValue();
+        }
+        return null;
+    }
+    
+    /**
+     *
+     * We are using the intersection of 2 sets here to make sure we only loop over edges that we will create. Previously we would loop over all the possible
+     * edges even if they were at different levels of nesting.
+     *
+     */
+    private long pairSubgroups(Set<String> commonKeys, EdgeDefinition edgeDef, RawRecordContainer event, String sourceGroup, String sinkGroup,
+                    TaskInputOutputContext<KEYIN,? extends RawRecordContainer,KEYOUT,VALUEOUT> context, ContextWriter<KEYOUT,VALUEOUT> contextWriter)
+                    throws IOException, InterruptedException {
+        
+        long edgesCreated = 0L;
+        
+        Multimap<String,NormalizedContentInterface> mSource = depthFirstList.get(edgeDef.getSourceFieldName());
+        Multimap<String,NormalizedContentInterface> mSink = depthFirstList.get(edgeDef.getSinkFieldName());
+        
+        // get the activity date from the event fields map
+        long activityDate = getActivityDate();
+        boolean validActivityDate = validateActivityDate(activityDate, event.getDate());
+        boolean activityEqualsEvent = false;
+        
+        // If the activity date is valid check to see if it is on the same day as the event date
+        if (validActivityDate) {
+            activityEqualsEvent = compareActivityAndEvent(activityDate, event.getDate());
+        }
+        
+        EdgeKey.DATE_TYPE date_type = getEdgeKeyDateType(validActivityDate, activityEqualsEvent, event.getDate());
+        
+        for (String subGroup : commonKeys) {
+            for (NormalizedContentInterface ifaceSource : mSource.get(subGroup)) {
+                for (NormalizedContentInterface ifaceSink : mSink.get(subGroup)) {
+                    EdgeDataBundle edgeValue = createEdge(edgeDef, event, ifaceSource, sourceGroup, subGroup, ifaceSink, sinkGroup, subGroup);
+                    if (edgeValue != null) {
+                        
+                        // have to write out the keys as the edge values are generated, so counters get updated
+                        // and the system doesn't timeout.
+                        edgesCreated += writeEdgesForDateType(edgeValue, context, contextWriter, date_type);
```

Comment:
Compared with the pre-refactored code, it looks like writeEdgesForDateType adds this code which was not here before.  I wonder if that was intentional.
```
         if (dateType.equals(EdgeKey.DATE_TYPE.ACTIVITY_ONLY)) {
            edgesCreated += writeEdges(value, context, contextWriter, EdgeKey.DATE_TYPE.EVENT_ONLY);
        }
```

Reply:
Yes.  I was trying to cut down on the number of arguments per function since it seemed to be out of control here (15 in some cases!).   We were repeatedly passing the two dates around and repeatedly checking their type.  The strategy for writing edges was dependent on the type, so I thought it clearer to extract some of that and update the method name to highlight this.
In the old writeEdges, it has ACTIVITY_ONLY and EVENT_ONLY written together thus:
`else {
                    edgesCreated += writeEdges(value, context, contextWriter, EdgeKey.DATE_TYPE.ACTIVITY_ONLY);
                    edgesCreated += writeEdges(value, context, contextWriter, EdgeKey.DATE_TYPE.EVENT_ONLY);`
The new code preserves this relationship in the snippet you pasted above.